hcl/fmt.go

```diff
@@ -151,3 +164,318 @@ func (f FormatResult) Formatted() string {
+		fmtBody(block.Body())
+	}
+}
+
+func fmtAttrExpr(tokens hclwrite.Tokens) hclwrite.Tokens {
+	logger := log.With().
+		Str("action", "hcl.fmtAttrExpr()").
+		Str("tokens", tokensStr(tokens)).
+		Logger()
+
+	logger.Trace().Msg("trimming newlines")
+
+	trimmed := trimNewlines(tokens)
+
+	logger = logger.With().
+		Str("trimmedTokens", tokensStr(trimmed)).
+		Logger()
+
+	// We are interested on lists, ignore the rest
+	if trimmed[0].Type == hclsyntax.TokenOBrack {
+
+		if isListComprehension(tokens) {
+			logger.Trace().Msg("list comprehension, ignoring")
+			return tokens
+		}
+
+		formattedList, pos := fmtListExpr(trimmed)
+		if pos != len(trimmed) {
+			panic(fmt.Errorf(
+				"last pos %d != tokens len %d for tokens: %q",
+				pos,
+				len(trimmed),
+				tokensStr(trimmed),
+			))
+		}
+		return formattedList
+	}
+
+	logger.Trace().Msg("not a list, returning tokens as is")
+	return trimmed
+}
+
+// fmtListExpr will adjust the given list tokens so they can be formatted
+// properly. It returns the adjusted tokens and the position of the first
+// token after the list ended.
+//
+// If there is no more tokens after the end of
+// the list the returned position will be equal to len(tokens).
+func fmtListExpr(tokens hclwrite.Tokens) (hclwrite.Tokens, int) {
+	logger := log.With().
+		Str("action", "hcl.fmtListExpr()").
+		Str("tokens", tokensStr(tokens)).
+		Logger()
+
+	logger.Trace().Msg("formatting list")
+
+	elemNextPos := 0
+	newTokens := hclwrite.Tokens{tokens[elemNextPos], newlineToken()}
+	elemNextPos++
+
+	skipNls := func() {
+		_, skipped := skipNewlines(tokens[elemNextPos:])
+		elemNextPos += skipped
+	}
+
+	for {
+		skipNls()
+
+		nextTokenType := tokens[elemNextPos].Type
+
+		if nextTokenType == hclsyntax.TokenComma {
+			elemNextPos++
+			continue
+		}
+
+		if nextTokenType == hclsyntax.TokenComment {
+			newTokens = append(newTokens, tokens[elemNextPos])
+			elemNextPos++
+			continue
+		}
+
+		if nextTokenType == hclsyntax.TokenCBrack {
+			logger.Trace().Msg("reached end of list")
+			break
+		}
+
+		logger.Trace().Msg("getting next element of the list")
+
+		element, nextPos := fmtNextElement(tokens[elemNextPos:])
+		elemNextPos += nextPos
+
+		logger.Trace().
+			Str("element", tokensStr(element)).
+			Str("tokens", tokensStr(tokens)).
+			Int("elemNextPos", elemNextPos).
+			Msg("new element got")
+
+		newTokens = append(newTokens, element...)
+		// Heredocs need to be handled differently, the comma must
+		// be on the next line in this case
+		if isHeredoc(element) {
+			newTokens = append(newTokens, newlineToken())
+		}
+		newTokens = append(newTokens, commaToken(), newlineToken())
+	}
+
+	newTokens = append(newTokens, closeBracketToken())
+	elemNextPos++
+
+	// Handling ["one"][0] and things like [[0]%[0]]
+	// We can also have newlines when dealing with operations
+
+	logger.Trace().Msg("checking if formatted list has operators/index access")
+
+	skipNls()
+
+	if elemNextPos == len(tokens) {
```

Comment:
maybe this check can be in the `skipNls()` and then it could return a `isEOF` bool.
We should return everytime skipNls() reaches the end but we are not checking this everytime. Not sure how to better handle this, a macro would be great =)

Reply:
hmm since we use this logic only here and most skipNls don't need to check it I'm not so sure. The code would be something like:

```
if skipNls() {
//handle EOF
}
```

Which is not extremelly intuitive IMHO (alternatives involving creating a isEOF var also feels kinda off to me) :thinking:

Reply:
reading past the end of the buffer is a common parser issue, that's why I pointed out that maybe it'd be useful, but yeah, if only here makes sense to check then dont bother adding this.